citiesapps_com: Update to citiesapp v2 api

lbloder opened the pull request:
Fixes #3460

Updates citiesapp related code to use v2 api.
Also updates wizard to be compatible.

@mampfes Thank you and all the contributors for your work on this integration. Its much appreciated. Also note that I am not a python developer, so any suggestions are welcome :)

5ila5 commented on custom_components/waste_collection_schedule/waste_collection_schedule/service/CitiesAppsCom.py:
```diff
@@ -1463,47 +1463,28 @@ def get_specific_citiy(self, search: str) -> dict:
+            if calendar["street"].lower().strip() == search.lower().strip():
                 return calendar
         raise SourceArgumentNotFoundWithSuggestions(
-            "calendar", [calendar["name"] for calendar in calendars]
+            "calendar", [calendar["street"] for calendar in calendars]
         )
```
Looks like the Excption call is wrong (from before you touched it) it should be 
```py
        raise SourceArgumentNotFoundWithSuggestions(
            "city", search, [city["name"] for city in cities]
        )
```
 in get_specific_citiy

and 
```py
        raise SourceArgumentNotFoundWithSuggestions(
            "calendar", search, [calendar["street"] for calendar in calendars]
        )
```
here.

lbloder commented:
Thanks for the feedback @5ila5, will look into it and update the PR.

overide6 commented:
I made the changed, but after reconfiguration, I still don't get the waste type in the calendar.
what do I need to do to get the waste types into the calendar and then setup the sensors again?
City: Str. Ruprecht an der Raab
Calendar: Unterfladnitz

thx

lbloder commented:
@overide6 Thank you for testing this. This seems to be the same issue as for Fürstenfeld, which I already have a fix for. In my initial code I did not handle the pagination, so Unterfladnitz is not returned from the API.

I will update the PR here once I'm done testing.

lbloder commented:
Short update from my side:
Looks like `Rudersdorf` is still using the v1 api. There is an api call which we can use to figure out which garbage collection version is used. I'm currently trying to figure out how to implement both v1 and v2 in a nice manner.

I'm not fully done yet, due to time constraints. I hope to have a working version by Monday.

@overide6 As a quick workaround you can replace:
```
https://api.v2.citiesapps.com/waste-management/by-city/{city_id}/areas
```

with

```
https://api.v2.citiesapps.com/waste-management/by-city/{city_id}/areas?pagination=limit:100
```

in `CitiesAppsCom.py`

overide6 commented:
@lbloder 
thank you, but I don't have this line in my `CitiesAppsCom.py` file located in the "service" folder
`https://api.v2.citiesapps.com/waste-management/by-city/{city_id}/areas`

could you tell me where to find it?
Thank you

lbloder commented:
@5ila5 
I reworked the `citiesapps_com` source and corresponding service to handle both v1 and v2 cities. It would be great if you could have another look. I tried to keep the separation between source and service, if you have suggestions for improvements, I'd be happy to adapt the code.

@overide6 maybe you could try the new implementation to see if that works for you.

overide6 commented:
@lbloder 
I now see the waste types in the calendar again, so it seems to work. thank you.

5ila5 commented on custom_components/waste_collection_schedule/waste_collection_schedule/service/CitiesAppsCom.py:
```diff
@@ -1459,39 +1459,42 @@ def get_specific_citiy(self, search: str) -> dict:
+            "data": api.fetch_garbage_plans(city_dict, calendar)
+        }
+
+    def get_garbage_calendars(self, city_id: str) -> list:
```
Type hinted as returning list, but returns `dict[str, bool|list]`

lbloder replied:
Ah, thank you, will update.

5ila5 commented:
The `Rudersdorf Rudersdorf 3` test case seems to fail. If this test case is not compatible with the api_v2, and it cannot be made to work, it's best to remove it and replace it with a working one

lbloder commented:
> The `Rudersdorf Rudersdorf 3` test case seems to fail. If this test case is not compatible with the api_v2, and it cannot be made to work, it's best to remove it and replace it with a working one

Looks like they migrated to the calendar v2 api in the meantime 🤦 . This was basically the reason I built it so we can handle both v1 and v2 :).
I'll keep the v1 compatibility in the code, not sure when they will migrate all cities to v2. Maybe I can find another city that still uses v1.
I'll update the test case.

lbloder commented:
Updated the Rudersdorf test case and added a suggestion to re-check the App if no calendar could be found.

Also, looks like they are migrating all cities to v2, as of this writing, there are only 7 cities left on v1. Did some spot checking and it seems like they don't return any up-to-date data. Guessing they will also be migrated to v2 when new data is available. 

Once all cities are on v2 we can probably remove all the v1 code.
